spec/java/util/stream/DoubleStream.main.lsl

```diff
@@ -0,0 +1,1216 @@
+    }
+
+
+    @Phantom proc sort_loop_outer (i: int, j: int, innerLimit: int): void
+    {
+        innerLimit = this.length - i - 1;
+        action LOOP_FOR(
+            j, 0, innerLimit, +1,
+            sort_loop_inner(j)
+        );
+    }
+
+
+    @Phantom proc sort_loop_inner (j: int): void
+    {
+        val idxA: int = j;
+        val idxB: int = j + 1;
+        val a: double = this.storage[idxA];
+        val b: double = this.storage[idxB];
+
+        if (a > b)
+        {
+            this.storage[idxA] = b;
+            this.storage[idxB] = a;
+        }
+    }
+
+
+    fun *.distinct (@target self: DoubleStream): DoubleStream
+    {
+        if (this.linkedOrConsumed)
+            _throwISE();
+
+        var distinctStorage: array<double> = null;
+        var distinctLength: int = 0;
+
+        val size: int = this.length;
+        if (size == 0)
+        {
+            distinctStorage = action ARRAY_NEW("double", 0);
+            distinctLength = 0;
+        }
+        else
+        {
+            val items: array<double> = this.storage;
+            action ASSUME(items != null);
+            action ASSUME(action ARRAY_SIZE(items) != 0);
+            action ASSUME(size == action ARRAY_SIZE(items));
+            var i: int = 0;
+            var j: int = 0;
+
+            // serialize stored items
+            val uniqueItems: list<Double> = action LIST_NEW();
+            val visited: map<Double, Double> = action MAP_NEW();
+            action LOOP_FOR(
+                i, 0, size, +1,
+                distinct_loopStoreItems(i, items, visited, j, uniqueItems)
+            );
+
+            // allocate space for unique items
+            distinctLength = j;
+            action ASSUME(distinctLength > 0);
+            action ASSUME(distinctLength <= size);
+            distinctStorage = action ARRAY_NEW("double", distinctLength);
+
+            // restore unique items back
+            action LOOP_FOR(
+                i, 0, distinctLength, +1,
+                distinct_loopRecoverItems(i, uniqueItems, distinctStorage)
+            );
+        }
+
+        // return a new instance
+        result = new DoubleStreamAutomaton(state = Initialized,
+            storage = distinctStorage,
+            length = distinctLength,
+            closeHandlers = this.closeHandlers,
+        );
+
+        this.linkedOrConsumed = true;
+    }
+
+    @Phantom proc distinct_loopStoreItems (i: int, items: array<double>, visited: map<Double, Double>, j: int, uniqueItems: list<Double>): void
+    {
+        val item: double = items[i];
+        if (action MAP_HAS_KEY(visited, item) == false)
+        {
+            action MAP_SET(visited, item, STREAM_VALUE);
+            action LIST_INSERT_AT(uniqueItems, j, item);
+            j += 1;
+        }
+    }
+
+    @Phantom proc distinct_loopRecoverItems (i: int, uniqueItems: list<Double>, distinctStorage: array<double>): void
+    {
+        distinctStorage[i] = action LIST_GET(uniqueItems, i);
```

Comment:
@dog-m As I see this code compiles (Idea didn't underline this line with red color)

Reply:
```
public interface SymbolicList<Value> {

    Value get(int index);
```
Type of value is "Double"

Reply:
Because we are boxing primitive values and **then** unboxing them, we can be sure *(and USVM should too)* that references before unboxing are valid (not `null`).
So, using the suggested sequence of operations we exclude 'null'-pointer check and make it clear that there will be a conversion taken place *(via `Double.doubleValue()` but it's not that important and will happen auto-magically by the Java compiler)*.

Reply:
Fixed;